### DISCUSSION TITLE
Is there a difference between these ways to define indexes?

### DISCUSSION THREAD

**Discussion by zolbayars** [prisma/prisma#21153] [Q&A] [answered]
### Question

**A**: 
```prisma
model clients {
  id                            Int                @id @unique(map: "id") @default(autoincrement()) @db.Int
  name                          String             @db.VarChar(255)
  domain                        String?            @db.VarChar(255) 

  @@index([name])
  @@index([domain])
}
```

**B**: 
```prisma
model clients {
  id                            Int                @id @unique(map: "id") @default(autoincrement()) @db.Int
  name                          String             @db.VarChar(255)
  domain                        String?            @db.VarChar(255) 

  @@index([name, domain])
}
```

Is there any difference between these two? 🤔

**Comment by ludralph** [comment, marked as answer]
Hi @zolbayars 👋 

Thank you for raising this question.

There is a difference between the two ways of defining indexes in your Prisma schema. You are creating two separate single-column indexes, one for the `name` field and another for the `domain` field. This means that each field will have its own index, which can speed up queries that filter or sort by either name or domain. The underlying SQL is shown below for the first `clients` model:
```sql
-- CreateTable
CREATE TABLE `clients` (
    `id` INTEGER NOT NULL AUTO_INCREMENT,
    `name` VARCHAR(255) NOT NULL,
    `domain` VARCHAR(255) NULL,

    UNIQUE INDEX `id`(`id`),
    INDEX `clients_name_idx`(`name`),
    INDEX `clients_domain_idx`(`domain`),
    PRIMARY KEY (`id`)
) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
```
1. The `INDEX clients_name_idx(name)` creates an index on the name column.  
2. The `INDEX clients_domain_idx(domain)` creates an index on the domain column, similar to the index on the name column.

Each index is associated with a specific column and is designed to improve the performance of queries involving that column. 

For the second `clients` model, you are creating a multi-column index on the `name` and `domain` fields. This means that the index is based on the combination of `name` and `domain`. This can speed up queries that filter or sort by both name and domain at the same time. However, this type of index may not be as efficient if you're only filtering or sorting by one of the fields. The underlying SQL is shown below:
```sql
CREATE TABLE `clients` (
    `id` INTEGER NOT NULL AUTO_INCREMENT,
    `name` VARCHAR(255) NOT NULL,
    `domain` VARCHAR(255) NULL,

    UNIQUE INDEX `id`(`id`),
    INDEX `clients_name_domain_idx`(`name`, `domain`),
    PRIMARY KEY (`id`)
) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
```
The `INDEX clients_name_domain_idx(name, domain)` is a composite index created on both the `name` and `domain` columns together. A composite index allows for efficient querying and sorting based on both `name` and `domain` columns simultaneously. It's useful when you need to search for rows that match specific combinations of `name` and `domain` values.

The choice between these two ways of defining indexes depends on the specific queries you expect to run on your data. If you often query by both name and domain, the multi-column index could be more beneficial. If you often query by either name or domain, but not usually both, the separate single-column indexes could be more beneficial.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by zolbayars** [reply]
Got it. Thanks for the detailed answer :) 
So that means I could define indexes like this as well? 
```
  @@index([name])
  @@index([domain])
  @@index([name, domain])
```

**Reply by janpio** [reply]
Yes.

In my experience, and some people are of different opinion, `@@index([name])` and `@@index([name, domain])` have the same effort for queries involving `name`, so I might leave off the first one. Danger of course is that you remove the composite index when you do not query for both any more, accidentally also hurting queries that only query by `name`. If you don't have a problem with space usage of indexes and with insert speed in that table, I hence recommend keeping both.
